### PR TITLE
feat: [#1075] Allows more discriminatedUnion values, use DiscriminatedValue as union element

### LINKED ISSUE #1075: Allow more types in discriminated unions
It seems like `z.discriminatedUnion` currently has quite narrow requirements for what you can use it with. Here are some example use cases that seem sound to me (in the sense that the discriminator should still be able to uniquely determine the union member) but currently aren't allowed:

```typescript
const A = z.object({ type: z.literal('a') });
const B = z.object({ type: z.literal('b') });
const C = z.object({ type: z.literal('c') });
const D = z.object({ type: z.literal('d').optional() });
const E = z.object({ type: z.enum(['e', 'ee', 'eee']) });

const AorB = z.discriminatedUnion('type', [A, B]);

// using another discriminated union with the same discriminator as a union member
const AorBorC = z.discriminatedUnion('type', [AorB, C]);

// the discriminator is optional in one union member
const CorD = z.discriminatedUnion('type', [C, D]);

// the discriminator is an enum
const CorE = z.discriminatedUnion('type', [C, E]);
```

Any chance these cases can be supported? It would great if the requirement for discriminated union members could be widened to something like "any type that has the discriminator as a field and the type of the discriminator is a subtype of `string | null | undefined` and is mutually exclusive between the discriminated union members", but supporting at least the cases I mentioned above would still be much appreciated.

### LINKED ISSUE #1884: discriminated union cant be nested
Currently it is impossible to allow discriminated unions as argument to another discriminated union.
Typescript however allows that so it should be allowed in zod as well

Example Typescript and zod:
https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgLzgXzgMyhEcDkyEAJvgLABQlMAnmAKZwCCcAvIpXHMcAM4DGUUMAB2AQxjQAjAC44MKAFd6Abk5wQvAOZzeC0VrUU0lanUYAhFuwTqwYqGJCy4IxSABG9KEa48BQiCiEtAATHKYYgA2vKrq-oLC4pJQLpEx9JQmVBS0DHAWFmwcFH58iUHJYXIKyr5w9o4g4a7uXj7x5YHBKWnRsVmmueYFxQAUVnAAPgUWAJRDeYwAwsUsMxZDlPwQInpYIuNgcstzbAB8JerAmHBjAIRgAHQJ3VWpZ7alXHA3d48vLpJEJQUKfdQ-OA7PYQKL0J5RCBaMbPRpOKQLb5cDD0DIlSFcaG8WHwxHI1EOJxgiHodQ4vFfSFEkkIpEop6aLSY7GDHJE+AAfWsKCeEA8ACt6PwYGNGXAoPRJdL6MQ5MgEcAYN5omNavQ5gAadSctVPPRCETIzFoTHbXb7AWTdjqsVKmVynj0ZaKGDLCB6XimtyebxjQ3qBWKWKqkVRTXaqJjdKxcPfBVulWmuNaxyJ5P6rK2ij8uCOorO0USqXu9QAdwAFhIg21Q6muJHo1n47ndUp9Ua04rq5nY92dfnrUWS47iurXsCtcQAKoiYC7Mb4Dsq-AGuAAbUdTF3ZYAukWAPTnuDeHBQXh2vaC1YV+eVCQqldrkQb9PD0i7g8j1LCwzyGS9fhETBvGAuBa2gABre9hnyAAtFDZyeUQoKgAAeJYIFuR1zkoIA

Here is the code for reference:

```ts
import { z } from 'zod'

type A = {
  discriminator1: true;
  msg: string;
}

type BA = {
  param1: number;
  discriminator2: false;
  discriminator1: false
}

type BB = {
  discriminator2: true;
  param2: number;
  discriminator1: false
}

type B = (BA | BB)

type C = A | B


const fn = (p: C) => {

  if (!p.discriminator1) {
    if (!p.discriminator2) {
      console.log(p.param1)
    } else {
      console.log(p.param2)
    }
  } else {
    console.log(p.msg)
  }
}

const _A = z.object({
  rejected: z.literal(true),
  msg: z.string()
})

const _BA = z.object({
  dieCutCosts: z.number(),
  reused: z.literal(false),
  rejected: z.literal(false)
})

const _BB = z.object({
  what: z.number(),
  reused: z.literal(true),
  rejected: z.literal(false)
})

const _B = z.discriminatedUnion('reused', [_BA, _BB])

// errors
const _C = z.discriminatedUnion('rejected', [_A, _B])

// infer _B works
type ZZ = z.infer<typeof _B>
```

### PR DESCRIPTION
**Pull request by maxArturo**
Partially closes https://github.com/colinhacks/zod/issues/1075 , fully closes #1884  with the following:

## feature: more valid discriminator values

The following are now well-supported discriminator values (enum/undefined could have been already)
- `z.undefined()`
- `z.null()`
- `z.enum()`
- `z.discriminatedUnion()`

This does **not** address functionality like the below, where a full primitive validator (like `z.string()`) is used as discriminator. This requires more changes and I'd like to address it separately.
```typescript
  const ResultSchema = z.discriminatedUnion("error", [
    z.object({ error: z.string().min(1) }), // error here
    z.object({ error: z.undefined(), value: z.string().min(1) }),
  ]);
```

## feature: use `ZodDiscriminatedUnion` as a union element

Allows adding a `ZodDiscriminatedUnion` as an element in the discriminated union, such that this is possible: 

```typescript
  const A = z.object({ type: z.literal("a") });
  const B = z.object({ type: z.literal("b") });
  const C = z.object({ type: z.literal("c").optional(), c: z.literal(true) });
  const AorB = z.discriminatedUnion("type", [A, B]);
  // using another discriminated union with the same discriminator as a union member
  const AorBorC = z.discriminatedUnion("type", [AorB, C]);
  type full = z.infer<typeof AorBorC>;
  /**
   * 
type full = {
    type: "a";
} | {
    type: "b";
} | {
    type?: "c" | undefined;
    c: true;
}
   */
```

Benchmarks for the new functionality (compared to two and four union elements, where `nested` is one of the options of `DiscriminatedUnion` itself):
```shell
$ yarn benchmark                                                                                                                                                                                           
yarn run v1.22.19
$ tsx src/benchmarks/index.ts
z.discriminatedUnion: double: valid: a x 994,112 ops/sec ±0.91% (85 runs sampled)
z.discriminatedUnion: double: valid: b x 964,468 ops/sec ±0.98% (85 runs sampled)
z.discriminatedUnion: double: invalid: null x 31,516 ops/sec ±1.10% (80 runs sampled)
z.discriminatedUnion: double: invalid: wrong shape x 34,323 ops/sec ±1.37% (81 runs sampled)
z.discriminatedUnion: many: valid: a x 948,977 ops/sec ±0.89% (87 runs sampled)
z.discriminatedUnion: many: valid: c x 944,571 ops/sec ±0.85% (85 runs sampled)
z.discriminatedUnion: many: invalid: null x 29,904 ops/sec ±1.24% (87 runs sampled)
z.discriminatedUnion: many: invalid: wrong shape x 32,622 ops/sec ±1.16% (88 runs sampled)
z.discriminatedUnion: nested: valid: a x 961,652 ops/sec ±0.82% (85 runs sampled)
z.discriminatedUnion: nested: valid: b x 963,830 ops/sec ±0.96% (86 runs sampled)
z.discriminatedUnion: nested: valid: c x 956,284 ops/sec ±0.88% (84 runs sampled)
z.discriminatedUnion: nested: invalid: null x 29,937 ops/sec ±1.19% (87 runs sampled)
z.discriminatedUnion: nested: invalid: wrong shape x 964,482 ops/sec ±0.91% (86 runs sampled)
✨  Done in 71.60s.
```
## feature: disjointed discriminators for nested DUs

Now the following is possible: 
```ts
  const subtype = z.discriminatedUnion("subtype", [
    z.object({
      type: z.literal("baz"),
      subtype: z.literal("able"),
    }),
    z.object({
      type: z.literal("bauble"),
      subtype: z.literal("beehive"),
      undertype: z.literal("alpha"),
    }),
    z.object({
      type: z.literal("baz"),
      subtype: z.literal("baker"),
    }),
  ]);

  const schema = z.discriminatedUnion("type", [
    z.object({
      type: z.literal("foo"),
    }),
    z.object({
      type: z.literal("bar"),
    }),
    subtype,
  ]);

/**
type schema = {
    type: "baz";
    subtype: "able";
} | {
    type: "bauble";
    subtype: "beehive";
    undertype: "alpha";
} | {
    type: "baz";
    subtype: "baker";
} | {
    type: "foo";
} | {
    type: "bar";
}
*/
```

Care was taken to ensure that there aren't any edge cases where nested DUs can be created without overlapping, resolvable discriminators at every level.

## Follow-ups
- Use the following zod validators as `DiscriminatedUnion` elements:
  - `z.intersection()`
  - `z.string()`
  - `z.boolean()`
  - others depending on implementation

### CODE REVIEW

**Comment by hornta** [comment]
Is this PR waiting for a review? I'm looking forward to this being merged.

**Comment by maxArturo** [comment]
@JacobWeisenburger do you have approval privileges? Do you mind looking at this? Thanks!

**Comment by JacobWeisenburger** [comment]
> @JacobWeisenburger do you have approval privileges? Do you mind looking at this? Thanks!

I can approve things, but I'm not sure I have enough experience to approve this. I really like the idea, I'm just not sure if there are unintended consequences.

**Comment by maxArturo** [comment]
@JacobWeisenburger understood! I tried to be thorough with the tests to catch any edge cases I could think of. With that said - this is additive so if anything, it should hopefully not impact any existing functionality. Can you think of anything else we can look at in order to merge? 

Maybe @colinhacks can chime in!

**Comment by teunmooij** [comment]
I'm working on #1768 , which adds ZodObject methods (like pick, omit) to the discriminated union. Hope this PR (which LGTM) can be merged before I create mine, because else my PR will be a breaking change for this one (and vice versa).

File: src/types.ts
Comment (teunmooij) [inline src/types.ts:2575]: What's the reason for adding the parentDiscriminator to `this._def`? It's not used anywhere and also it's a unwanted side-effect. 
Consider ZodDiscriminatedUnion `A` that is added to another ZodDiscriminatedUnion `B`. Now when `A` is used directly it wrongfully has `B` as parent discriminator. Also when later added to `C` it now has 2 parent discriminators.

File: src/__tests__/discriminatedUnions.test.ts
Comment (teunmooij) [inline src/__tests__/discriminatedUnions.test.ts:201]: Would it make sense to either:
- not allow adding discriminated unions with same discriminator as option, but add a merge function instead or
- in the implementation of create flatten the structure and add the individual sub-schema's to the options if the discriminator is the same

File: src/types.ts
Comment (maxArturo) [reply src/types.ts:2575]: Hey @teunmooij thanks for your feedback! I think this was leftover when I was thinking of how to solve this earlier (pre disjointed discriminators). For some reason I can't recall I thought it was useful; but yeah no need for this at the moment. The method can just recurse without having to store the parent discriminators around.

File: src/__tests__/discriminatedUnions.test.ts
Comment (maxArturo) [reply src/__tests__/discriminatedUnions.test.ts:201]: Oh great idea @teunmooij! I suspect it might be easier on users to just have this happen behind the scenes on `create()`. Let me add that functionality.

File: src/types.ts
Comment (maxArturo) [reply src/types.ts]: resolved with [223b34b](https://github.com/colinhacks/zod/pull/1589/commits/223be4b0cc835b3352ed795335d6ff8f55f2d3ed)

**Comment by Fuzzyma** [comment]
Hey @maxArturo can I be of any help getting this pr landed? I am looking for a solution to nested discriminated unions for month already and your pr looks like the most complete effort :)

**Comment by maxArturo** [comment]
@Fuzzyma I just have one PR comment to resolve which should happen today or tomorrow (traveling ATM). Otherwise it should be good to go. 

The main concern is getting someone with merger role to approve or give FB; @scotttrinh @colinhacks anyone else come to mind @JacobWeisenburger ?

**Comment by Fuzzyma** [comment]
Sounds awesome! Hope your notebook battery doesn't die and you can work 😁

Maybe their discord server has anyone who can approve?

File: src/__tests__/discriminatedUnions.test.ts
Comment (maxArturo) [reply src/__tests__/discriminatedUnions.test.ts:201]: @teunmooij see [latest commit](https://github.com/colinhacks/zod/pull/1589/commits/0f16bf71779e02a94fe81b5dc7648834cc6522ab)

**Comment by maxArturo** [comment]
OK! hopefully more 👀  on this can get it to the finish line. Let me know what else I can do to get this approved!

File: src/__tests__/discriminatedUnions.test.ts
Comment (teunmooij) [reply src/__tests__/discriminatedUnions.test.ts:201]: @maxArturo , LGTM. Nice improvement. (if it was up to me I'd pull your branch to double check and then merge it, would be nice to have this functionality in Zod. It's taking a very very long time for some reason)

**Review by teunmooij — APPROVED** [review]

**Comment by Fuzzyma** [comment]
Sooo, who is allowed to merge this now? :D

**Comment by maxArturo** [comment]
@Fuzzyma I think Scott and Colin. Here's what I got from @scotttrinh late last week on discord: 

> I don't have a ton of time these days to weigh in on changes like this, tbh. I can try to take some time soon, but honestly we've fallen back to just waiting for Colin to have time to swing around and do a big round of reviews and merging.

I think until then, we'll have to sit tight and wait for these to get pulled into actual package releases. Also, if you're desperate for this you can always install directly from a github repo into your package.json: 

    git+https://<github>/<org or person>/<repo>.git#<branch>

**Comment by Fuzzyma** [comment]
@maxArturo that is actually what I ended up doing end I found that this it doesn't work be sure the fist files are (rightfully) not in the repo. But you can get around that by adding a prepack script that builds the package. So I had to fork the fork to make that change and install from my fork :D

Also: this package gets more and more traction nowadays and the prs will only get more.
Maybe more people should get merge rights (after proving their capabilities). Any plans on this?

**Comment by helmturner** [comment]
> @maxArturo that is actually what I ended up doing end I found that it doesn't work because the dist files are (rightfully) not in the repo. But you can get around that by adding a prepack script that builds the package. So I had to fork the fork to make that change and install from my fork :D
> 
> Also: this package gets more and more traction nowadays (because it's awesome!) and the prs will only get more. Maybe more people should get merge rights (after proving their capabilities). Any plans on this?

I second this... Zod is an amazing contribution to the TS community, and there are a lot of excellent contributions in the pipeline. It would be a shame to see that pipeline bottlenecked by a busy BDFL.

**Comment by peterfoldes** [comment]
This is an amazing PR, just what I would need. Looking forward to this making into a release! 👍

**Comment by hornta** [comment]
@joseph-lozano Do you want to take a look at this and maybe have it merged? @maxArturo There are some conflicts in the branch.

**Comment by teunmooij** [comment]
Wow this is taking long the get merged...

**Comment by scotttrinh** [comment]
Hey friends and community members, just wanted to check in here with the latest. Thanks so much to everyone's hard work for getting this PR in shape, it's well done and I appreciate that lots of people from across the community have contributed time and thought to it. Sincerely, thanks!

Aside from my own availability for staying on top of reviews and PRs here, I've been a little shy about building atop the existing `discriminatedUnion` work without getting more alignment and direction from @colinhacks . Since the appearance of this feature (which I helped merge without a lot of alignment 😅 ), he's expressed feelings that there should be a better way to achieve this with a different API. Turns out he has had a chance to take a swing at that in #2106 and I think we're unlikely to merge any improvements to `discriminatedUnion` until we get a more firm resolution there.

Apologies for not addressing these dynamics sooner, and I hope people have been able to find ways around the current limitations or maybe used their own fork while the dust is (slowly) settling on this. `discriminatedUnion` was originally merged as a "better than nothing" solution to the most simple cases, so one workaround to any issues is to use a bare `union` which _should_ work for any cases we're not explicitly handling here. I know the error reporting is really noisy for `union` though, so I sympathize with wanting something better here.

Thanks for your patience and understanding! Let's keep this open and we'll continue to gather feedback on the proposed changes in #2106 and leave the discussion here for technical or design questions about this change specifically.

**Comment by scotttrinh** [comment]
I think there is a separate issue that's been raised here about governance and how we maintain such a popular library, and I'd welcome a lively discussion about this over on the Discord if anyone has the inclination or experience to help us out with that.

**Comment by tianhuil** [comment]
Would love to see this @maxArturo!

**Comment by maxArturo** [comment]
Hey yall. Sorry for having been AFK all these months - lots of personal things were happening. 

In short - I'm either happy to cleanup/fix this PR or close it, depending on whatever happens with https://github.com/colinhacks/zod/issues/2106. Since we'd just need this functionality to be integrated in whatever way the repo owners seem fit, we're basically on a holding pattern there. 

I will say too - a bit of the slowness on these overarching API decisions did cause me a bit to lose focus and switch gears on other tasks. I figure that's the case with other contributors too. But that's another story and no justification to leave people hanging, so I apologize again.

There's also https://github.com/teunmooij/zod-discriminated-union which @teunmooij put together in case you need this _now_ 😄.

**Comment by divmgl** [comment]
Are there any major blockers to merging this PR? Can't currently nest discriminated unions so this is a blocker for me. Any workarounds would be nice.

**Comment by colinhacks** [comment]
I appreciate the work done for this PR

The first comment in this thread is a little out of date, as all of the below have supported as discriminators for a long time now:

- `z.undefined()`
- `z.null()`
- `z.enum()`
- `z.nativeEnum()`

Other PRs were merged around the same time this one was created to support these things. This PR adds support for `ZodOptional` as well, which I think is good, but perhaps not a highly demanded feature.

---

The main thrust of this PR is the ability to nest ZodDiscriminatedUnions.

My position on this is that it adds to much implementational complexity relative to the value of this API. It's a very niche feature that would increase Zod's unzipped bundle size by ~2kb. Immediately that isn't worth it.

On top of that, this can be achieved quite simply with `.options` and destructuring. I'm surprised this wasn't suggested earlier. 

```ts
const A = z.discriminatedUnion([ ... ]);
const B = z.discriminatedUnion([ ...A.options, ...more options])
```

On top of *that* ZodDiscriminatedUnion will probably not exist in its current form in the next major version of Zod for the reasons described [here](https://github.com/colinhacks/zod/issues/2106). 

---

I've just merged this PR that adds support for more types (but not `ZodDiscriminatedUnion`): https://github.com/colinhacks/zod/pull/3346

@maxArturo I really appreciate the work you put into this. It's a beautiful approach. 🙌

**Comment by maxArturo** [comment]
Thanks @colinhacks for taking the time to explain your reasoning. That makes sense to me and I think it’s a good and clean API. 

Given that it’s one of my first contributions to this repo, I was really surprised by the traction it received by many users of your library! The DU-ish functionality seems to be in high demand. 

Do you have plans on adding the `z.switch()` functionality in an upcoming release? I’d love to give it a shot if we know this is the path forward - it doesn’t _seem_ too hard which is a good indication that it’s a good solution 😋.

**Comment by colinhacks** [comment]
> Do you have plans on adding the z.switch() functionality

@maxArturo 

This would land in Zod 4, if at all. I'm doing a long-overdue house cleaning right now before starting on Zod 4. I have the advantage of working full-time on Zod for the next couple months.

At the moment though, I'm leaning toward consolidating everything into `z.union()`. When a ZodUnion instance is created, it can iterate over its elements and find some fast-and-easy "qualification checks" for each element. When parsing is happening, ZodUnion can run these fast checks for each union element to quickly disqualify certain options before running a full parse. 

Put more generally, ZodUnion can get a lot fancier in terms of runtime optimization, making `z.discriminatedUnion` irrelevant. Ultimately, it will be up to the user to craft a `ZodUnion` such that the inferred type is a proper discriminated union.

**Comment by sladiri** [comment]
Just my 2c as a new user of Zod: I found nested discriminated unions intriguing, allows me to have domain logic in types: For example a CRUD form object with or without an id property (create/update), but then nested are the "actual" discriminated unions, certain properties depending on the object kind.

I have used the forks https://github.com/teunmooij/zod-discriminated-union and https://github.com/parisholley/zod-discriminated-union and it nesting is one major feature I see actually.

**Comment by Freddis** [comment]
Let me just leave here a workaround for zod3. We can actually create artificial descriminator field, which provides all the possible freedom and you don't have to touch the  original data.

I'm saying zod 3 since i was working with zod 3, assuming that zod 4 can descriminate over undefined.

```typescript
import { discriminatedUnion, literal, object, preprocess, type TypeOf } from "zod";

import { PayloadAttributesSchema } from "./AirShoppingRSPayloadAttributes";
import { DataListsSchema } from "./DataLists";
import { DistributionChainSchema } from "./DistributionChain";
import { ErrorResponseSchema } from "./ErrorResponse";
import { OffersGroupSchema } from "./OffersGroup";

const responseSchema = object({
  HasError: literal(false),
  Response: object({
    DataLists: DataListsSchema,
    OffersGroup: OffersGroupSchema,
  }),
  DistributionChain: DistributionChainSchema,
  PayloadAttributes: PayloadAttributesSchema,
});

//todo: switch to zod4 and use discriminatedUnion with undefined
// export const IATA_AirShoppingRS_BodySchema = union([ErrorResponseSchema, responseSchema]);
export const IATA_AirShoppingRS_BodySchema = preprocess((val: unknown) => {
  if(typeof val !== 'object' || !val) {
    return val;
  }
  const record: Record<string, unknown> = {
    ...val,
    HasError: 'Error' in val,
  };
  return record;
},discriminatedUnion('HasError',[ErrorResponseSchema.extend({HasError: literal(true)}), responseSchema]));
export type IATA_AirShoppingRS_Body = TypeOf<typeof IATA_AirShoppingRS_BodySchema>;
`

**Pull request closed without merging**
